src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/UpdateNuGetConfigPackageSourcesMappings.cs

```diff
@@ -34,45 +37,126 @@ public class UpdateNuGetConfigPackageSourcesMappings : Task
         /// </summary>
         public string[] SourceBuildSources { get; set; }
 
+        public string VmrRoot { get; set; }
+
+        private string SBRPCacheSourceName = "source-build-reference-package-cache";
+
         public override bool Execute()
         {
             string xml = File.ReadAllText(NuGetConfigFile);
             string newLineChars = FileUtilities.DetectNewLineChars(xml);
             XDocument document = XDocument.Parse(xml);
             XElement pkgSrcMappingElement = document.Root.Descendants().FirstOrDefault(e => e.Name == "packageSourceMapping");
+            XElement pkgSourcesElement = document.Root.Descendants().FirstOrDefault(e => e.Name == "packageSources");
+            if (pkgSourcesElement == null)
+            {
+                return true;
+            }
 
             if (pkgSrcMappingElement == null)
             {
-                return true;
+                pkgSrcMappingElement = new XElement("packageSourceMapping");
+                document.Root.Add(pkgSrcMappingElement);
             }
 
-            // Union all package sources to get the distinct list.  These will get added to the source-build sources.
-            string[] packagePatterns = pkgSrcMappingElement.Descendants()
-                .Where(e => e.Name == "packageSource")
-                .SelectMany(e => e.Descendants().Where(e => e.Name == "package"))
-                .Select(e => e.Attribute("pattern").Value)
-                .Distinct()
-                .ToArray();
+            Hashtable allSourcesPackages = new Hashtable();
+            Hashtable currentPackages = new Hashtable();
+            Hashtable referencePackages = new Hashtable();
+            Hashtable previouslyBuiltPackages = new Hashtable();
+            Hashtable oldSourceMappingPatterns = new Hashtable();
```

Comment:
We usually recommend to use the generic version of Hashtable<T> to avoid all the boxing from object to the concrete type.

Reply:
Yeah, that's a better option - will update, thanks!

Reply:
Fixed with https://github.com/dotnet/installer/pull/19114/commits/3f02bc95f3815f096425785f6fd3e2a351888a7e